addon/adapters/osf-adapter.js

```diff
@@ -18,97 +24,199 @@ export default DS.JSONAPIAdapter.extend(HasManyQuery.RESTAdapterMixin, DataAdapt
             url += '/';
         }
         return url;
     },
     /**
-     * Construct a URL for a relationship create/update/delete. Has the same
-     * signature as buildURL, with the addition of a 'relationship' param
+     * Construct a URL for a relationship create/update/delete.
      *
      * @method _buildRelationshipURL
+     * @param {DS.Snapshot} snapshot
      * @param {String} relationship the relationship to build a url for
      * @return {String} a URL
      **/
-    _buildRelationshipURL(modelName, id, snapshot, requestType, query, relationship) { // jshint ignore:line
+    _buildRelationshipURL(snapshot, relationship) {
         var links = relationship ? snapshot.record.get(
             `relationshipLinks.${Ember.String.underscore(relationship)}.links`
         ) : false;
         if (links && (links.self || links.related)) {
             return links.self ? links.self.href : links.related.href;
+        }
+        return null;
+    },
+    _createRelated(store, snapshot, createdSnapshots, relationship, url, isBulk = false) { // jshint ignore:line
+        // TODO support bulk create?
+        // if (isBulk) {
+        //
+        // }
+        if (createdSnapshots.record) {
+            return createdSnapshots.record.save({
+                adapterOptions: {
+                    nested: true,
+                    url: url
+                }
+            });
         } else {
-            return this.buildURL(...arguments);
+            return createdSnapshots.map(s => s.record.save({
+                adapterOptions: {
+                    nested: true,
+                    url: url
+                }
+            }));
         }
     },
-    _serializeHasMany(serialized) {
-        if (serialized.length > 1) {
-            serialized = {
-                data: serialized.map(function(record) {
-                    var data = record.data;
-                    return data;
-                })
-            };
+    _updateRelated(store, snapshot, updatedSnapshots, relationship, url, isBulk = false) {
+        return this._doRelatedRequest(store, snapshot, updatedSnapshots, relationship, url, 'PATCH', isBulk);
+    },
+    _addRelated(store, snapshot, addedSnapshots, relationship, url, isBulk = false) {
+        return this._doRelatedRequest(store, snapshot, addedSnapshots, relationship, url, 'POST', isBulk);
+    },
+    _removeRelated(store, snapshot, removedSnapshots, relationship, url, isBulk = false) {
+        return this._doRelatedRequest(store, snapshot, removedSnapshots, relationship, url, 'DELETE', isBulk).then(response => response || []);
+    },
+    _deleteRelated(store, snapshot, removedSnapshots) { // jshint ignore:line
+        return this._removeRelated(...arguments).then(() => {
+            if (removedSnapshots.record) {
+                removedSnapshots = [removedSnapshots];
+            }
+            removedSnapshots.forEach(s => s.record.unloadRecord());
```

Comment:
Previously we'd avoided direct store manipulation. To help me read this code, can you please give me an example of the intended behavior that this is designed to improve upon?

How are "remove" and "delete" intended to be different?